File: 0264/solution.java
class Solution {
    public int nthUglyNumber(int n) {
        int[] uglyNums = new int[n];
        uglyNums[0] = 1;
        int twos = 0;
        int threes = 0;
        int fives = 0;

        for (int i = 1; i < n; i++) {
            uglyNums[i] = Math.min(uglyNums[twos] * 2, Math.min(uglyNums[threes] * 3, uglyNums[fives] * 5));
            if (uglyNums[i] == uglyNums[twos] * 2) {
                twos++;
            }
            if (uglyNums[i] == uglyNums[threes] * 3) {
                threes++;
            }
            if (uglyNums[i] == uglyNums[fives] * 5) {
                fives++;
            }
        }
        return uglyNums[n - 1];
    }
}

// time O(n)
// space O(n)

File: 0264/solution.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        two, three, five = 0, 0, 0
        nums = [0] * n
        nums[0] = 1

        for i in range(1, n):
            next_two = nums[two] * 2
            next_three = nums[three] * 3
            next_five = nums[five] * 5

            nums[i] = min(next_two, next_three, next_five)

            if nums[i] == next_two:
                two += 1
            if nums[i] == next_three:
                three += 1
            if nums[i] == next_five:
                five += 1

        return nums[-1]
